oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,370 @@
+
+Rationale
+*********
+
+The majority of the concerns expressed in the Motivation section revolve around a lack of shared dependencies and the way in which MFEs are currently siloed from each other, preventing us from creating a more seamless, cohesive experience.
+
+Module federation specifically addresses these use cases exactly. It's right-sized to the problem at hand, can be accomplished with a minimum of impact on our existing MFEs, and can be done in a backwards compatible way (more on that below).
+
+Backward Compatibility
+**********************
+
+We intend to maintain backwards compatibility while migrating to using module federation. We can do this by creating a separate set of Webpack configurations in ``@edx/frontend-build`` and separate build targets in converted MFEs; the footprint of module federation on "guest" MFEs is very small, requiring virtually no code changes in the application itself, and a few additional configuration options in the MFE's Webpack config to identify exposed components.
+
+Ultimately MFEs will no longer be responsible for initializing frontend-platform or rendering the header and footer.  We will follow the DEPR process for retiring this code in MFEs once (and if) we make the module federation architecture required.
+
+In the interim, MFEs will have both a webpack configuration that exposes modules for consumption by other hosts as alternate entry points (to use Webpack parlance) _and_ the primary entry point which initializes frontend platform and loads the header/footer.  Early POC testing indicates this won't be a problem.
+
+Reference Implementation
+************************
+
+A proof of concept has already been created demonstrating how Webpack module federation works with two Open edX micro-frontends based on the frontend-template-application. The POC has several shared libraries (``react``, ``react-dom``, and ``@openedx/paragon``), and loads a React component module from a guest MFE into the page of a host MFE. It supports hot module replacement during development, and runs on the two MFEs' `webpack-dev-server` instances.
+
+The POC repositories can be found here:
+
+- `frontend-app-mf-host <https://github.com/davidjoy/frontend-app-mf-host>`_
+- `frontend-app-mf-guest <https://github.com/davidjoy/frontend-app-mf-guest>`_
+
+To convert this into a reference implementation, we need to minimally:
+
+- Create a new "shell" micro-frontend to be the top-level "host" for all our other micro-frontends.
+- Create module federation-based development and production Webpack configurations in ``@edx/frontend-build``.
+- Modify the Webpack configuration to share the complete list of shared dependencies from the shell.
+- Pick an existing MFE (or two) to convert to use module federation. Add build targets to these "guest" micro-frontends that can be used to build them in module-federation mode.
+- Extend the Webpack configuration in the MFEs by defining what modules each "guest" MFE exposes.
```

Comment:
If at all possible, I would prefer that the "list of modules exposed by each MFE" be done using the `package.json` `exports` field, which is the standard NodeJs way of doing this. And then it's not webpack-specific.

Reply:
Looking at the format of `exports` I think we could do this.  It _appears_ to be essentially the same as the `exposes` config in ModuleFederationPlugin, in which case we could just import package.json into the webpack config and read the `exports` field.

Reply:
Doc reference for those following along at home: https://nodejs.org/api/packages.html#subpath-exports

Reply:
I've added this to the reference implementation description. 👍🏻